crates/sui-source-validation/src/lib.rs

```diff
@@ -136,38 +183,93 @@ impl<'a> BytecodeSourceVerifier<'a> {
             .rpc_client
             .get_object(obj_id)
             .await
-            .map_err(DependencyVerificationError::DependencyObjectReadFailure)?;
+            .map_err(SourceVerificationError::DependencyObjectReadFailure)?;
 
         let obj = obj_read
             .object()
-            .map_err(DependencyVerificationError::SuiObjectRefFailure)?;
+            .map_err(SourceVerificationError::SuiObjectRefFailure)?;
 
         match obj.data.clone() {
             SuiRawData::Package(pkg) => Ok(pkg),
             SuiRawData::MoveObject(move_obj) => Err(
-                DependencyVerificationError::ObjectFoundWhenPackageExpected(obj_id, move_obj),
+                SourceVerificationError::ObjectFoundWhenPackageExpected(obj_id, move_obj),
             ),
         }
     }
 }
 
-fn get_module_bytes_map(compiled_package: &CompiledPackage) -> ModuleBytesMap {
+fn get_module_bytes_map(
+    compiled_package: &CompiledPackage,
+    include_deps: bool,
+    root_address: Option<AccountAddress>,
+) -> Result<ModuleBytesMap, SourceVerificationError> {
     let mut map: ModuleBytesMap = HashMap::new();
-    for (package, local_unit) in &compiled_package.deps_compiled_units {
-        let CompiledUnitEnum::Module(m) = &local_unit.unit else {
-            continue;
-        };
 
-        let module = m.name;
-        let address = m.address.into_inner();
+    #[allow(clippy::type_complexity)]
+    fn make_map_entry(
```

Comment:
If I understand what this is doing, it's replacing the zero address in the root module with the new root address, in order to support comparing source where the address is `0x0` with on-chain bytecode (where it won't be), and it will also complain if the `self_handle` address for a root module is *not* `0x0`.

If that's the case, then I think this is back-to-front, and making it the right way around should simplify a lot of things:

- In the case where we are verifying the root compiled units (i.e. the package itself and not (necessarily) its dependencies), we should assume that the source we have already has the correct address assigned for the package (i.e. not `0x0`) and we should be happy with verification failing on the package if that address is `0x0`, because it means the author hasn't identified an address that it has been published at.
- This means that we don't need to supply an explicit root address, we can extract it from its compiled module (from the self handle), and complain if that is `0x0`.
- This also gets rid of a newly introduced class of errors, which is nice, and simplifies this code (which should be able to go back to how it was before).

Reply:
> we should assume that the source we have already has the correct address assigned for the package

hmm my thought process was that this would likely be run on existing source code published to sites like GH, in which case the source address is usually 0x0.
For example [Sui 8192](https://github.com/EthosWallet/Sui8192/blob/main/move/Move.toml) and [MoveMate](https://github.com/pentagonxyz/movemate/blob/main/sui/Move.toml).

In this case a user flow would be for Sui 8192:
1. Clone the source form the repo
2. Run the verif command against the source and specify the on-chain address

In your approach, one would have to clone, then modify the address in the toml file before running verif.
1. Clone the source form the repo
2. Modify the toml file with on-chain address
3. Run the verif command against the source and specify the on-chain address

Please correct me if I'm wrong.

Reply:
It's intended for existing source code that people have published to GitHub, but it's supposed to be used to verify that you can depend on the package at a given address.

If the package source is published with the address at 0x0, then even if you know an address that it is published at, you would not be able to use it at that address without forking the package and updating the addresses on it.  The current way to depend on that package would be to re-publish it with your own, which does not require source verification.

This is something we want to change as part of  automatic address management, but that may not be something we can get to in the medium term, so my suggestion is for us to implement the simpler case now (no need to substitute package addresses) because that is the one that is definitely needed as part of a flow of verifying that you can depend on a package on-chain.

Unfortunately, devnet packages are not going to be a great example of this way of working, because of how often we wipe the network -- packages don't get stable addresses to put into their source code.